src/gui/torrentcontentmodel.cpp

```diff
@@ -510,20 +544,11 @@ void TorrentContentModel::setupModelData(const BitTorrent::AbstractFileStorage &
 
         for (const QStringView pathPart : asConst(pathFolders))
         {
-            const QString folderPath = pathPart.toString();
-            TorrentContentModelFolder *newParent = currentParent->childFolderWithName(folderPath);
-            if (!newParent)
-            {
-                newParent = new TorrentContentModelFolder(folderPath, currentParent);
-                currentParent->appendChild(newParent);
-            }
-            currentParent = newParent;
+            const QString folderName = pathPart.toString();
+            currentParent = currentParent->findChildFolderOrInsert(folderName);
         }
         // Actually create the file
-        TorrentContentModelFile *fileItem = new TorrentContentModelFile(
-                    Utils::Fs::fileName(info.filePath(i)), info.fileSize(i), currentParent, i);
-        currentParent->appendChild(fileItem);
-        m_filesIndex.push_back(fileItem);
+        m_filesIndex.push_back(currentParent->insertFile(Utils::Fs::fileName(info.filePath(i)), info.fileSize(i), i));
```

Comment:
too long and too much action on a line. please move `currentParent->insertFile()` to a variable.

Reply:
I would even change `insertFile()` to accept `TorrentContentModelFile *` instead.

Reply:
> I would even change insertFile() to accept TorrentContentModelFile * instead.

then I've to expose FolderAdaptor::m_folder, also this way it mirrors the folder insertion.

Reply:
Otherwise you can handle "file index" by FolderAdaptor (pass reference in constructor and store it like m_folder) and add item in `FolderAdaptor::insertItem()`

Reply:
> Otherwise you can handle "file index" by FolderAdaptor (pass reference in constructor and store it like m_folder) and add item in FolderAdaptor::insertFile()

Reply:
> > Otherwise you can handle "file index" by FolderAdaptor (pass reference in constructor and store it like m_folder) and add item in FolderAdaptor::insertFile()
> 
> sorry, I don't understand.

https://github.com/qbittorrent/qBittorrent/pull/16183/files#r789430431